core/src/main/scala/org/apache/spark/scheduler/ShuffleMapTask.scala

```diff
@@ -58,8 +59,18 @@ private[spark] class ShuffleMapTask(
     // Deserialize the RDD using the broadcast variable.
     val deserializeStartTime = System.currentTimeMillis()
     val ser = SparkEnv.get.closureSerializer.newInstance()
+    // Kryo deserialization is not thread-safe w.r.t. underlying buffer
+    // create a copy of the buffer if Kryo is being used
+    val copy = if (SparkEnv.get.closureSerializer.isInstanceOf[KryoSerializer]) {
```

Comment:
I suppose that there's some code duplication between ShuffleMapTask and ResultTask, but maybe that's not easily avoidable.  We might be able to factor the deserialization code into a method in `Task` and have it be parametrized by the type passed to `ser.deserialize`. Maybe you could try doing this to see if it's possible, since it would be way cleaner?